### ISSUE TITLE
About the working mode during testing

### ISSUE THREAD

**Issue by jefferson814** [Feliciaxyao/NavMorph#8]
Hello,

Thank you for your inspiring work on the VLN-CE. We have successfully reproduced the experimental results for your VLN-3DFF version. However, we have a question regarding the working mode of the world model during the testing phase.

In the implementation details section of your paper's supplementary materials, you mention that the world model's foresight action planner helps refine the final navigation decision logits during the online testing phase.

We noticed that this part of the logic has been commented out in the provided code (see ’ss_trainer_ETP.py‘, lines 1548-1553). When we re-enabled this code and ran the provided model, we observed a performance degradation.

Here are our results on r2r val-unseen:

Without the foresight action planner (consistent with the results in the paper):

2025-09-24 09:50:48,796 Episodes evaluated: 1839
2025-09-24 09:50:48,796 Average episode steps_taken: 231.428494
2025-09-24 09:50:48,796 Average episode distance_to_goal: 5.811098
2025-09-24 09:50:48,797 Average episode success: 0.475258
2025-09-24 09:50:48,797 Average episode oracle_success: 0.568244
2025-09-24 09:50:48,797 Average episode path_length: 22.275305
2025-09-24 09:50:48,797 Average episode collisions: 0.273976
2025-09-24 09:50:48,797 Average episode spl: 0.330320
2025-09-24 09:50:48,797 Average episode ndtw: 0.447660
2025-09-24 09:50:48,797 Average episode sdtw: 0.324713
2025-09-24 09:50:48,797 Average episode ghost_cnt: 36.591626

With the foresight action planner enabled:

2025-09-24 16:01:15,916 Episodes evaluated: 1839
2025-09-24 16:01:15,916 Average episode steps_taken: 214.845568
2025-09-24 16:01:15,916 Average episode distance_to_goal: 5.808981
2025-09-24 16:01:15,916 Average episode success: 0.463839
2025-09-24 16:01:15,916 Average episode oracle_success: 0.548668
2025-09-24 16:01:15,916 Average episode path_length: 20.349816
2025-09-24 16:01:15,916 Average episode collisions: 0.265670
2025-09-24 16:01:15,916 Average episode spl: 0.329917
2025-09-24 16:01:15,917 Average episode ndtw: 0.458796
2025-09-24 16:01:15,917 Average episode sdtw: 0.322258
2025-09-24 16:01:15,917 Average episode ghost_cnt: 35.779772

From our experiments, it appears that only the CEM plays a core role during the testing phase. Could you please clarify how the world model is intended to be used during testing to improve the agent's decision-making?

I look forward to your reply, and I would be very grateful if you could respond.

**Comment by Feliciaxyao** [comment]
Thank you for the detailed question. Below is a clarification regarding how the world model and the foresight action planner are intended to function during testing.

---

### 1. About the foresight action planner described in the supplementary material

As described in the supplementary materials, the foresight action planner can be used to predict future waypoints and slightly adjust the final action logits by giving higher scores to candidate viewpoints that are closer to the predicted waypoint.

In our internal experiments, this refinement produced only very limited gains. This suggests that the predicted waypoint and the latent dynamics learned by the world model are already highly consistent, which confirms that the world model successfully captures the environment dynamics.

Importantly, enabling this refinement is not required for the world model to function during testing.

---

### 2. Why this part of the code is commented out

The commented logic reflects a design choice in the released version. The public codebase currently provides an inference pipeline that relies only on the navigation policy learned during world-model–guided training, without using the optional foresight-based refinement.

We apologize for any confusion this may have caused. You may re-enable this mechanism, but as your results show, its performance benefit is not very significant. This is also one of the directions we plan to further improve.

---

### 3. Disabling the foresight planner does not disable the world model or CEM

The foresight planner is not the world model itself.

* Disabling it only removes the auxiliary inference-time heuristic that adjusts logits based on predicted waypoints.
* The Contextual Evolution Memory (CEM) and the latent world model remain fully active.
* As discussed in the paper, CEM plays a dual role:

  1. providing additional predictive signals, and
  2. enabling the agent’s self-evolving capability during navigation.

Therefore, disabling the foresight planner does not affect the core functionality of the world model.
